### PR TITLE
Added Warning and Exception annotations

### PR DESCRIPTION
**Pull request by BvB93**
See #54.

Added annotations and tests for:
* ~~``np.warnings``~~ (removed as of https://github.com/numpy/numpy-stubs/pull/57#discussion_r410731135)
* ``np.ModuleDeprecationWarning``
* ``np.VisibleDeprecationWarning``
* ``np.ComplexWarning``
* ``np.RankWarning``
* ``np.TooHardError``
* ``np.AxisError``

### CODE REVIEW

File: numpy-stubs/__init__.pyi [hunk 1]
```diff
@@ -778,3 +779,17 @@ trunc: ufunc
 
 # TODO(shoyer): remove when the full numpy namespace is defined
 def __getattr__(name: str) -> Any: ...
+
+# Warnings
+warnings = _warnings
```
Comment (person142) [inline numpy-stubs/__init__.pyi:784]: Har, I think we should leave out warnings here and declare that any code using `np.warnings` is officially broken.

File: numpy-stubs/__init__.pyi [hunk 2]
```diff
@@ -778,3 +779,17 @@ trunc: ufunc
...
+class ModuleDeprecationWarning(DeprecationWarning): ...
+class VisibleDeprecationWarning(UserWarning): ...
+class ComplexWarning(RuntimeWarning): ...
+class RankWarning(UserWarning): ...
+
+# Errors
+class TooHardError(RuntimeError): ...
+class AxisError(ValueError, IndexError):
+    # Note that all parameters for __init__ (except self) can technically be Any.
+    # So it's a question here whether or not we should allow what's possible or what's intended
```
Comment (person142) [inline numpy-stubs/__init__.pyi:794]: The way it is done here seems right to me.

File: tests/pass/warnings_and_errors.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ComplexWarning = np.ComplexWarning  # type: Type[RuntimeWarning]
+RankWarning = np.RankWarning  # type: Type[UserWarning]
+
+TooHardError = np.TooHardError  # type: Type[RuntimeError]
+AxisError1 = np.AxisError  # type: Type[ValueError]
+AxisError2 = np.AxisError  # type: Type[IndexError]
+
+np.AxisError(1)
+np.AxisError(1, ndim=2)
+np.AxisError(1, ndim=2, msg_prefix='error')
```
Comment (person142) [inline tests/pass/warnings_and_errors.py:18]: Maybe add some cases in `tests/fail` for the constructor.

File: tests/pass/warnings_and_errors.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import types
+from typing import Type
+
+import numpy as np
+
+warnings = np.warnings  # type: types.ModuleType
+ModuleDeprecationWarning = np.ModuleDeprecationWarning  # type: Type[DeprecationWarning]
```
Comment (person142) [inline tests/pass/warnings_and_errors.py:7]: I would move these to `test/reveal/warnings_and_errors.py` and do something like

```
reveal_type(ModuleDeprecationWarning())  # E: numpy.DeprecationWarning
...
```

File: numpy-stubs/__init__.pyi (hunk 2 above)
Comment (person142) [reply numpy-stubs/__init__.pyi:794]: (Which means we can delete the comment I think.)

**Comment by BvB93** [comment]
Thanks for feedback! I've addressed your points in the latest push.

**Comment by person142** [comment]
Great, thanks @BvB93.

**Pull request merged**